### PR TITLE
boards: add support for HiFive Unleashed

### PR DESCRIPTION
**Pull request by katsuster**
This patch adds new support for SiFive HiFive Unleashed which has
SiFive FU540 SoC, DDR and some peripherals.

First version is minimum only using UART, SPI and DDR memory area.

### CODE REVIEW

**Review by galak — COMMENTED** [review]
Is the Unleashed supported in qemu?

**Comment by katsuster** [comment]
Yes, QEMU supported a part of Unleashed hardware and memory map.
But not supported all (Ex. SPI is not exist on QEMU).

```
qemu-system-riscv64 -machine sifive_u,dumpdtb=sifive_u.dtb
dtc sifive_u.dtb -I dtb -O dts -o sifive_u.dts
```

[sifive_u.dts.txt](https://github.com/zephyrproject-rtos/zephyr/files/6632297/sifive_u.dts.txt)

**Comment by carlescufi** [comment]
@katsuster I noticed that this board is discontinued. Is there another board in production with the same SoC?

**Comment by katsuster** [comment]
@carlescufi Yes, already discontinued. And unfortunately I don't know another boards has FU540.
SiFive is moving next generation SoC FU740 (HiFive Unmatched).

If this support is not valuable for Zephyr, please drop it.

**Comment by katsuster** [comment]
@galak Sorry, I got a mistake. QEMU emulates SPI and GPIO of Unleashed. And can boot from Flash.
https://qemu.readthedocs.io/en/latest/system/riscv/sifive_u.html

**Comment by katsuster** [comment]
@galak @carlescufi 
Sorry for ask again... If Zephyr project policy does not welcome the patch for a discontinued board as Unleashed, I will close this PR.
But if not, I want to progress this PR and continue to add supported device (or add driver if we need).

**Comment by cfriedt** [comment]
@katsuster - I'm not sure what zephyr policy is w.r.t. unsupported boards, but do you have access to an Unmatched? I'm sure a PR for that board would be quite welcome and likely some of the same code could be recycled.

**Comment by cwshu** [comment]
@cfriedt @katsuster 
According to their dtsi in linux kernel, I think most peripherals in Unleashed/Unmatched are highly compatible.
Use uart driver as example: 
- Unmatched compatible string is  "sifive,fu740-c000-uart", "sifive,uart0"
- Unleashed compatible string is "sifive,fu540-c000-uart", "sifive,uart0"

And spi/i2c/gpio/pwm/eth are the same as uart.

**Comment by katsuster** [comment]
@cfriedt
Yes, I also have HiFive Unmatched and want to send PR for Unmatched in next PR. So if this PR dropped, I copy and modify codes for next PR. But if accepted, I just add few codes and DTS for supporting Unmatched.
@cwshu 
Thanks a lot for Linux information. I think we can reuse almost same driver codes for Unleashed and Unmatched :)

**Comment by katsuster** [comment]
I found a issue of current OpenOCD that does not support large size SPI Flash so I'm trying to fix and upstream it.
Currently we cannot burn the Zephyr flash ROM image into SPI Flash of Unleashed and Unmatched.

After that I'll try to create patch for HiFive Unmatched.

**Comment by cfriedt** [comment]
@katsuster - Could you add support for both in this PR? I think that would sway anyone who might be hesitant to accept

**Review by cfriedt — APPROVED** [review]
I'm approving on the condition that the Unmatched can be added in this PR :-)

**Comment by fkokosinski** [comment]
Hi! I managed to get your code working in emulation on Renode (see attached [patch](https://github.com/zephyrproject-rtos/zephyr/files/6968386/0001-boards-run-HiFive-Unleashed-on-Renode.patch.txt)). I had to rebase your branch onto upstream/main, though. Maybe it would be beneficial to run this board in simulation on Renode, as it supports Ethernet/QSPI/I²C/PWM for this board?

File: dts/riscv/riscv64-fu540.dtsi [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	#address-cells = <1>;
+	#size-cells = <1>;
+	compatible = "SiFive,FU540-C000", "fu540-dev", "sifive-dev";
```
[the remark this answers is not in the document]
Comment (katsuster) [reply dts/riscv/riscv64-fu540.dtsi:12]: Thanks! I will send other PR to fix FE310's same mistake.

[the remark this answers is not in the document]
Comment (katsuster) [reply dts/riscv/riscv64-fu540.dtsi]: Done.

**Comment by katsuster** [comment]
Add support for HiFive Unmatched.
And I will check build error of SPI flash sample.

**Comment by katsuster** [comment]
@galak @carlescufi @cfriedt
I added HiFive Unmatched, it's currently available board.
Unleashed is discontinued but as @fkokosinski said this is useful on Renode environment (and also on QEMU?).

**Comment by cfriedt** [comment]
I would love to merge both / all 😉

File: boards/riscv/hifive_unleashed/hifive_unleashed_defconfig [hunk 2]
```diff
@@ -0,0 +1,16 @@
+CONFIG_SOC_SERIES_RISCV_SIFIVE_FREEDOM=y
+CONFIG_SOC_RISCV_SIFIVE_FU540=y
+CONFIG_BOARD_HIFIVE_UNLEASHED=y
+CONFIG_RISCV_MACHINE_TIMER=y
+CONFIG_PLIC=y
+CONFIG_PINMUX=y
```
Comment (fkokosinski) [inline boards/riscv/hifive_unleashed/hifive_unleashed_defconfig:6]: I'm getting the following warning when building either `hifive_unleashed` or `hifive_unmatched`:
```
CMake Warning at ../../CMakeLists.txt:718 (message):
  No SOURCES given to Zephyr library: drivers__pinmux

  Excluding target from build.
```

Is this intentional?

Comment (katsuster) [reply boards/riscv/hifive_unleashed/hifive_unleashed_defconfig:6]: Oh, not intentional. I'll fix it. Thanks!

Comment (katsuster) [reply boards/riscv/hifive_unleashed/hifive_unleashed_defconfig]: Fixed.

**Review by fkokosinski — APPROVED** [review]
LGTM

**Review by nashif — APPROVED** [review]

**Pull request merged**